### DISCUSSION TITLE
Using with Typescript

### DISCUSSION THREAD

**Discussion by ballcoach12** [hyperjump-io/json-schema#33] [Q&A]
My apologies if this question is off topic.

I am following the example on the main readme.md.   When I import as in the example code in a unit test  `ts` file:

`import { addSchema, validate,  } from '@hyperjump/json-schema/draft-2020-12';`

I get the following error:

`SyntaxError: Cannot use import statement outside a module`

I recognize that this is related to different module formats, but I'm not exactly sure how to resolve it.  I have tried making modifications to my `package.json` file and to my `tsconfig.json` file, but I still can't get past the error.

**Comment by ballcoach12** [comment]
As a follow-up, there seems to  be some issue with how Jest compiles the code.   I have the following configuration in `jest.config.js`, which works for other dependencies that I import, but not for json-schema:

```
module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  transform: {
    '^.+\\.ts?$': 'ts-jest',
  },
  transformIgnorePatterns: ['<rootDir>/node_modules/'],
};

**Comment by jdesrosiers** [comment]
This is really a TypeScript configuration issue, which is not something I have a lot of knowledge about, but I'll do my best to help.

Here's the setup that I use for running one-off TS code with `npx tsx myscript.ts`. This configuration compiles the code to ESM so it can get executed by a node project configured to use ESM.

`tsconfig.json`
```json
{
  "compilerOptions": {
    "module": "es2020",
    "target": "es2015",
    "moduleResolution": "Node16",
    "esModuleInterop": true,
    "strict": true
  }
}
```

I don't remember what all of those things do, but iirc, `"moduleResolution": "Node16"` is the magic incantation that I was missing when I was in your position trying to get all the pieces to line up properly.

`package.json`
```json
{
  ...
  "type": "module",
  ...
}
```

This sets the node project to use ESM instead of the default CommonJS. If you want/need your node project to use CommonJS instead, I think you'd need to remove `"type": "module"` and change `"module"` to `"commonjs"`. But, that's just a guess. I don't have a working example to say for sure.

**Comment by jdesrosiers** [comment]
For mocha I had to [configure](https://github.com/hyperjump-io/json-schema/blob/main/.mocharc.json) a "loader" specifically for supporting ESM libraries. There might be some similar configuration for Jest.

**Comment by GabenGar** [comment]
If your code runs through a bundler (aka any transpilation tool which is not typesript), then you should try running with `"moduleResolution": "bundler"` first. This might or might not fix your specific problem, but worth a try when shooting in the dark.
